### DISCUSSION TITLE
Allow duplicate device names within same site/cluster

### DISCUSSION THREAD

**Discussion by AliMickey** [netbox-community/netbox#17913] [Help Wanted!]
I have defined a cluster to represent a hypervisor, within which there will be multiple VMs.

If there are two different VMs both named "test", the second will fail validation when adding due to `Virtual machine name must be unique per cluster.`.

~~This behaviour does not match ordinary device names being duplicated (where the device ID is appended to the name).~~

I know this can be worked around by defining a custom field and setting the true name there, however I prefer consistency with the existing 'normal' device objects.

Thanks

**Comment by candlerb** [comment]
> This behaviour does not match ordinary device names being duplicated (where the device ID is appended to the name).

As far as I know, there is no case where "the device ID is appended to the name". Can you give an example of a series of steps which demonstrates this behaviour?

Both devices and VMs have name uniqueness constraints. They are slightly different:

* For devices, the combination of (lowercased device name, site ID, tenant ID) must be unique ([code](https://github.com/netbox-community/netbox/blob/v4.1.6/netbox/dcim/models/devices.py#L789-L790))
* For virtual machines, the combination of (lowercased VM name, cluster ID, tenant ID) must be unique ([code](https://github.com/netbox-community/netbox/blob/v4.1.6/netbox/virtualization/models/virtualmachines.py#L158-L159))

If you want two different VMs with the same name in the same cluster (which would be confusing), then either:
1. Assign those different VMs to different tenants
2. Replace the VM name with something that *is* unique (e.g. in a Proxmox cluster there's the numeric VM ID)

**Reply by AliMickey** [reply]
Hi, 

Appreciate your response, I have attempted to reproduce that behaviour and have been unsuccessful. I mixed it up in my memory, it happens when a device name is left blank, as shown in the screenshot:

<img width="233" alt="image" src="https://github.com/user-attachments/assets/4f0dc2ca-c201-470f-a1a8-fbe75bf12089">

That being said, is there any technical reason for enforcing uniqueness within sites/clusters (without using tenants)? All routes already use ids anyway, and I can understand it may be confusing to track duplicate devices from a user standpoint, but there are valid use cases for supporting duplicate names.

Thanks again.

**Reply by candlerb** [reply]
> That being said, is there any technical reason for enforcing uniqueness within sites/clusters

Only that names are supposed to usable to identify something, and if they are not unique, they can't be used for this purpose (and are likely to be a source of confusion: e.g. "please delete the VM called 'test'")

Device names can be left blank, because devices are also used for dumb rack hardware like blanking panels and shelves, and it would be annoying having to give each one a distinct name.

If you describe your virtualization platform, perhaps someone else who uses that platform can comment on how they deal with naming.  If the platform lets you assign the name "test" to two different VMs, then presumably there must be some other unique ID it uses to distinguish between them.

**Reply by AliMickey** [reply]
Thanks again for your reply.

Maybe my use case is not as common as I would have thought for Netbox. 

Basically, I have multiple hypervisors (clusters) in VirtFusion with multiple VMs for various customers. My intention is to sync basic VM information into Netbox (name, IP addresses, cluster) primarily for IPAM. 

The dilemma is that there is a low chance of customers using common names for servers within the same hypervisor (test, pve, vps1), resulting in the original 'issue' that I raised. VirtFusion does also allow customers to define a separate field 'hostname' (which must be a valid FQDN) that would further reduce the chance or even eliminate duplicate names being synced to Netbox, however hostname is an optional field which would result in some VMs without names being synced.

Regarding your suggestion to use tenants, VirtFusion allows duplicate server names per user so not a solution, unfortunately.

If my use case sounds unreasonable and "non-standard" for the primary purpose of Netbox's DCIM, I'm happy to fallback to using a custom field instead as a workaround.
